cloudinit/config/cloud-init-schema.json

```diff
@@ -543,6 +543,244 @@
+            },
+            "ssl_dir": {
+              "type": "string",
+              "description": "The path to the puppet SSL directory. Default depends on ``install_type``"
+            },
+            "csr_attributes_path": {
+              "type": "string",
+              "description": "The path to the puppet csr attributes file. Default depends on ``install_type``"
+            },
+            "package_name": {
+              "type": "string",
+              "description": "Name of the package to install if ``install_type`` is ``packages``. Default: ``puppet``"
+            },
+            "exec": {
+              "type": "boolean",
+              "default": false,
+              "description": "Whether or not to run puppet after configuration finishes. A single manual run can be triggered by setting ``exec`` to ``true``, and additional arguments can be passed to ``puppet agent`` via the ``exec_args`` key (by default the agent will execute with the ``--test`` flag). Default: ``false``"
+            },
+            "exec_args": {
+              "type": "array",
+              "description": "A list of arguments to pass to 'puppet agent' if 'exec' is true Default: ``['--test']``",
+              "items": {
+                "type": "string"
+              }
+            },
+            "start_service": {
+              "type": "boolean",
+              "default": true,
+              "description": "By default, the puppet service will be automatically enabled after installation and set to automatically start on boot. To override this in favor of manual puppet execution set ``start_service`` to ``false``"
+            },
+            "conf": {
+              "type": "object",
+              "description": "Every key present in the conf object will be added to puppet.conf. As such, section names should be one of: ``main``, ``server``, ``agent`` or ``user`` and keys should be valid puppet configuration options. The configuration is specified as a dictionary containing high-level ``<section>`` keys and lists of ``<key>=<value>`` pairs within each section. The ``certname`` key supports string substitutions for ``%i`` and ``%f``, corresponding to the instance id and fqdn of the machine respectively.\n\n``ca_cert`` is a special case. It won't be added to puppet.conf. It holds the puppetserver certificate in pem format. It should be a multi-line string (using the | yaml notation for multi-line strings).",
+              "additionalProperties": false,
+              "properties": {
+                "main": {
+                  "type": "object"
+                },
+                "server": {
+                  "type": "object"
+                },
+                "agent": {
+                  "type": "object"
+                },
+                "user": {
+                  "type": "object"
+                },
+                "ca_cert": {
+                  "type": "string"
```

Comment:
Where are we grabbing these top-level section definitions? I agree we don't want to boil the ocean and define schema for al params that puppet provides, but how do we know that we aren't missing another config section. Docs I've seen look sparse on this as I couldn't find a [ca_cert] section

Reply:
I'm taking this from the existing documentation:
"section names should be one of: ``main``, ``server``, ``agent`` or
``user`` and keys should be valid puppet configuration options."
https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_puppet.py#L46-L56

https://puppet.com/docs/puppet/7/config_file_main.html#config_file_main-puppet-conf-config-sections also lists 4 config sections.

`ca_cert` I'm stealing directly from the code as current documentation looks to be wrong:
https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_puppet.py#L282

Reply:
That's good enough for me!